proposals/001_resource_kubernetes_patch.md

```diff
@@ -0,0 +1,165 @@
+- The resource's metadata (i.e name and namespace) 
+- The type of patch operation 
+- The patch itself 
+
+In the kubernetes provider, it would look like this. 
+
+```hcl
+resource "kubernetes_patch" {
+  kind = "namespace"
+
+  metadata {
+    name = "default"
+  }
+ 
+  patch = jsonencode({
+    metadata = {
+      labels = {
+        test = "test"
+      }
+    }
+  })
+ 
+  type = "merge"
+}
+```
+
+In the kubernetes-alpha provider:
+
+```hcl
+resource "kubernetes_patch" {
+  provider = kubernetes-alpha
+
+  kind = "namespace"
+
+  metadata {
+    name = "default"
+  }
+ 
+  patch = {
+    metadata = {
+      labels = {
+        test = "test"
+      }
+    }
+  }
+ 
+  type = "merge"
+}
+```
+
+## Possible Alternatives
+
+A possible alternative to the generic patch resource above is to implement use-case specific resources for managing things like annotations, tolerations, and so on. 
+
+For example:
+
+```hcl
+resource "kubernetes_annotations" {
+  kind = "pod"
+
+  metadata {
+    name = "pod-name"
+  }  
+
+  annotations = {
+    "test-annotation" = "value"
+  }
+}
+```
+
+The big downside of this is increased maintenance burden – we will have to implement a new resource for every use case that merits a resource being patched.  
+
+
+## Discussion
+
+
+1. **Should we implement this in the kubernetes provider, or in the kubernetes-alpha provider?**
+
+Arguments _for_ kubernetes provider:
+- Stability – users of the main provider won’t need to add kubernetes-alpha which still in flux. 
+- Simple patch doesn’t have as many exotic challenges as custom resources.
+- Easier implementation than in the alpha provider because it uses the main Terraform SDK. 
+
+Arguments _against_ kubernetes provider:
+- The patch has to be stored as a text string and used with `jsondecode`/`jsonencode`.
+- We will have to do JSON munging to do any validation on the patch.
+
+
+
+Arguments _for_ kubernetes-alpha provider:
+- Cleaner syntax, no need for `jsonencode()`.
+- Adding a highly wanted feature that incentivizes people to adopt the kubernetes-alpha provider.
+- Code already exists in kubernetes-alpha to do things like validate against the OpenAPI schema at plan time.
+
+Arguments _against_ kubernetes-alpha provider:
+- Scope of the alpha provider becomes less tightly circumscribed around custom resources
+- More conservative users will have to wait for the alpha provider to hit GA 
+
+
+2. **Should we support the JSON patch type of just strategic and merge?**
+
+The k8s documentation actually doesn’t illustrate the JSON patch type – it’s only there in the `--help` information.
+
+You can do everything with “merge” and “strategic” that json patch can do (I think). None of the commenters in the issue seem to have asked for this and the examples were strategic patches.
+
+
+3. **Should the resource be a “logical resource” that is applied once and then considered created forever, or should we read the resource to see if the patch needs to be applied again?**
```

Comment:
I think reading the resource would be a better user experience than a one-time create. A logical resource would only make sense when you're creating an object that doesn't need to exist on the API server (such as the ephemeral CertificateSigningRequest or generating a TLS key locally and storing the result in state).

But for a patch resource, users would expect that their changes remain in place (on the server) even after the initial terraform apply, because terraform tracks the changes of all the other things they're applying.

This would help us to maintain the CRUD model a bit better. We would just be applying CRUD to a part of the resource, but at least that part of the resource would fit into the Terraform model of declarative configuration.